### ISSUE TITLE
Don't loop forever when trying to create a directory

### ISSUE THREAD

**Issue by dpoeschl** [dotnet/roslyn#8133]
Ported from http://roslyn.codeplex.com/workitem/357 by @slaks

My decompiler found the following code in MetadataAsSourceFileService.GetGeneratedFileAsync:

```
            while (!Directory.Exists(var_15))
            {
                try
                {
                    Directory.CreateDirectory(var_15);
                }
                catch (DirectoryNotFoundException)
                {
                }
                catch (UnauthorizedAccessException)
                {
                }
            }
```

This seems like a bad idea; the loop should be bounded.

**Comment by SLaks** [comment]
Current source (and reason):
http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs,116

**Comment by aochagavia** [comment]
I am interested in fixing this. How many retries should we allow? Would it be ok to have a counter and try it 100 times for example?

**Comment by paul1956** [comment]
Make sure you give up the CPU and put in some delay. In VB I use application.doevents, not sure what C#  equivalent is. You could also check system time in the loop and exit after a reasonable time interval.

**Comment by aochagavia** [comment]
What about something like:

```c#
int tries = 0;
while (!Directory.Exists(var_15))
{
    if (tries > 10)
        throw ...; // Which exception would be suitable here?
    if (tries > 0)
        Thread.Sleep(50);

    tries++;
    try
    {
        Directory.CreateDirectory(var_15);
    }
    catch (DirectoryNotFoundException)
    {
    }
    catch (UnauthorizedAccessException)
    {
    }
}
```

**Comment by NicolasDorier** [comment]
For information, I stumbled upon this issue because my visual studio was hanging because of something related.

The hanging visual studio was caught into a loop. Procmon was showing me "PENDING DELETE" operations. When I tried to go to the folder with windows explorer, I was rejected as access denied, even as administrator.

With procexp, I found the culprit was another visual studio instance taking a lock on some files inside the folder, when I closed all my visual studio, the MetadataAsSource disappeared.
